solver/src/solver/matcha_solver.rs

```diff
@@ -0,0 +1,439 @@
+        let mut client = Box::new(MockMatchaApi::new());
+        let mut allowance_fetcher = Box::new(MockAllowanceManaging::new());
+
+        let sell_token = H160::from_low_u64_be(1);
+        let buy_token = H160::from_low_u64_be(1);
+
+        let allowance_target = shared::addr!("def1c0ded9bec7f1a1670819833240f027b25eff");
+        client.expect_get_swap().returning(move|_| {
+            Ok(SwapResponse {
+            sell_amount: U256::from_dec_str("100").unwrap(),
+             buy_amount: U256::from_dec_str("91").unwrap(),
+             allowance_target,
+            price: 0.91_f64,
+            to: shared::addr!("def1c0ded9bec7f1a1670819833240f027b25eff"),
+            data: web3::types::Bytes(hex::decode(
+                "d9627aa40000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000016345785d8a00000000000000000000000000000000000000000000000000001206e6c0056936e100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc20000000000000000000000006810e776880c02933d47db1b9fc05908e5386b96869584cd0000000000000000000000001000000000000000000000000000000000000011000000000000000000000000000000000000000000000092415e982f60d431ba"
+            ).unwrap()),
+            value: U256::from_dec_str("0").unwrap(),
+        })});
+
+        allowance_fetcher
+            .expect_get_approval()
+            .times(1)
+            .with(eq(sell_token), eq(allowance_target), eq(U256::from(100)))
+            .returning(move |_, _, _| {
+                Ok(Approval::Approve {
+                    token: sell_token,
+                    spender: allowance_target,
+                })
+            });
+
+        let solver = MatchaSolver {
+            client,
+            allowance_fetcher,
+        };
+
+        let order_passing_limit = LimitOrder {
+            sell_token,
+            buy_token,
+            sell_amount: 100.into(),
+            buy_amount: 90.into(),
+            kind: model::order::OrderKind::Sell,
+            ..Default::default()
+        };
+        let order_violating_limit = LimitOrder {
+            sell_token,
+            buy_token,
+            sell_amount: 100.into(),
+            buy_amount: 110.into(),
+            kind: model::order::OrderKind::Sell,
+            ..Default::default()
+        };
+
+        let result = solver
+            .settle_order(order_passing_limit)
+            .await
+            .unwrap()
+            .unwrap();
+        assert_eq!(
+            result.clearing_prices(),
+            &hashmap! {
+                sell_token => 91.into(),
+                buy_token => 100.into(),
+            }
+        );
+
+        let result = solver.settle_order(order_violating_limit).await.unwrap();
+        assert!(result.is_none());
+    }
+    #[tokio::test]
+    async fn test_satisfies_limit_price_for_buy_order() {
```

Comment:
nit: I'd be for combining the two limit order tests (buy & sell) side if that leads to significantly less code.